### PR TITLE
Too many notifications and logs when offheap threshold is reached

### PR DESCRIPTION
**Pull request by mathieucarbou**

### CODE REVIEW

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java [hunk 1]
```diff
@@ -105,22 +107,36 @@ public boolean reserve(long size) throws IllegalArgumentException {
...
+      newT = 0;
+    }
+    if (threshold.compareAndSet(curT, newT)) {
+      if (newT > curT) {
+        // increase from 0->75 or 75->90
+        LOGGER.info(MESSAGE_PROPERTIES.getProperty(newT == 90 ? OFFHEAP_WARN_KEY : OFFHEAP_INFO_KEY), identifier, percentOccupied);
+        onReservationThresholdReached.accept(this, new ThresholdChange(curT, newT));
+      } else if (newT < curT) {
+        // decrease from 90->75 or 75->0
+        if(newT == 75) {
```
Comment (GaryWKeim) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: formatting

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java [hunk 2]
```diff
@@ -61,10 +61,11 @@ public OffHeapResourcesProvider(OffheapResourcesType configuration) {
...
-      OffHeapResourceImpl offHeapResource = new OffHeapResourceImpl(identifier.getName(), size, (res, percent) -> {
+      OffHeapResourceImpl offHeapResource = new OffHeapResourceImpl(identifier.getName(), size, (res, update) -> {
         for (EntityManagementRegistry registry : registries) {
           Map<String, String> attrs = new HashMap<>();
-          attrs.put("percentOccupied", String.valueOf(percent));
+          attrs.put("oldThreshold", String.valueOf(update.old));
+          attrs.put("threshold", String.valueOf(update.now));
           attrs.put("capacity", String.valueOf(res.capacity()));
           attrs.put("available", String.valueOf(res.available()));
           registry.pushServerEntityNotification(res.getManagementBinding(), "OFFHEAP_RESOURCE_THRESHOLD_REACHED", attrs);
```
Comment (GaryWKeim) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java]: Will the EventLogMessageListener handler for OFFHEAP_RESOURCE_THRESHOLD_REACHED do anything with threshold or is that only for tests?

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java [hunk 3]
```diff
@@ -105,22 +107,36 @@ public boolean reserve(long size) throws IllegalArgumentException {
...
-      LOGGER.info(MESSAGE_PROPERTIES.getProperty(OFFHEAP_INFO_KEY), identifier, percentOccupied);
-      onReservationThresholdReached.accept(this, 75);
+      newT = 75;
+    } else {
+      newT = 0;
+    }
+    if (threshold.compareAndSet(curT, newT)) {
+      if (newT > curT) {
+        // increase from 0->75 or 75->90
+        LOGGER.info(MESSAGE_PROPERTIES.getProperty(newT == 90 ? OFFHEAP_WARN_KEY : OFFHEAP_INFO_KEY), identifier, percentOccupied);
```
Comment (anthonydahanne) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: used to be warn - why the change ?

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java [hunk 4]
```diff
@@ -105,22 +107,36 @@ public boolean reserve(long size) throws IllegalArgumentException {
...
+        onReservationThresholdReached.accept(this, new ThresholdChange(curT, newT));
+      } else if (newT < curT) {
+        // decrease from 90->75 or 75->0
+        if(newT == 75) {
+          LOGGER.info(MESSAGE_PROPERTIES.getProperty(OFFHEAP_INFO_KEY), identifier, percentOccupied);
```
Comment (anthonydahanne) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: same thing, why the change in log level ?

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java (hunk 2 above)
Comment (mathieucarbou) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourcesProvider.java]: EventLogMessageListener will need to use these 2 values to check if it is an increase or decrease and which threshold are now reached if you want to display alerts or just warnings

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java (hunk 3 above)
Comment (mathieucarbou) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: no it was info for 75% and warn for 90% so no change here

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java (hunk 4 above)
Comment (mathieucarbou) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: again, no change: 75% was mapped to info level by @ljacomet

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java (hunk 3 above)
Comment (anthonydahanne) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: ?

Comment (anthonydahanne) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: ah you tried to handle both cases (75 and 90 in the same LOGGER)

Comment (mathieucarbou) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: before: 

```
if(percentOccupied >= 90) { 
    LOGGER.warn(MESSAGE_PROPERTIES.getProperty(OFFHEAP_WARN_KEY), identifier, percentOccupied); 
}} else if (percentOccupied >= 75) {
    LOGGER.info(MESSAGE_PROPERTIES.getProperty(OFFHEAP_INFO_KEY), identifier, percentOccupied);
}
```

OFFHEAP_INFO_KEY != OFFHEAP_WARN_KEY ;-)

Comment (anthonydahanne) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: now I believe the behavior is changed to : 
```
if(percentOccupied >= 90) { 
    LOGGER.info(MESSAGE_PROPERTIES.getProperty(OFFHEAP_WARN_KEY), identifier, percentOccupied); 
}} else if (percentOccupied >= 75) {
    LOGGER.info(MESSAGE_PROPERTIES.getProperty(OFFHEAP_INFO_KEY), identifier, percentOccupied);
}
```

Comment (mathieucarbou) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: ahhh you were talking about the logger method called!

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java [hunk 5]
```diff
@@ -105,22 +107,40 @@ public boolean reserve(long size) throws IllegalArgumentException {
...
-      LOGGER.info(MESSAGE_PROPERTIES.getProperty(OFFHEAP_INFO_KEY), identifier, percentOccupied);
-      onReservationThresholdReached.accept(this, 75);
+      newT = 75;
+    } else {
+      newT = 0;
+    }
+    if (threshold.compareAndSet(curT, newT)) {
+      if (newT > curT) {
+        // increase from 0->75 or 75->90
+        if(newT == 90) {
```
Comment (GaryWKeim) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: formatting

Comment (mathieucarbou) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapResourceImpl.java]: this whole file is not formatted properly so i always forgot to format only the updated parts because a cmd+shif+f would change the whole file sadly.

I wish i could just reformat the whole file - would be easier.

**Comment by GaryWKeim** [comment]
LGTM
